### PR TITLE
nixos/systemd-boot: init boot counting

### PR DESCRIPTION
**Pull request by JulienMalka**
## Description of changes

New and reworked version of #273062. #273062 had the problem of being unable to work with past entries that it was trying to scan because they were missing an "sort-key" item. I have updated it to make fewer assumption on what is an entry present on disk (it now only expects `linux` and `initrd` which are mandatory items).
Tested on one of my machines.
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by r-vdp** [comment]
@JulienMalka in #283547 I proposed to use the `sort-key` to allow to reliably sort non-NixOS entries after NixOS ones, to avoid situations where a headless server boots into memtest86 for instance. It seems that the default entry is not always correctly selected otherwise and the first entry ends up being selected.

Do you see a way to support that in combination with this PR?

**Comment by JulienMalka** [comment]
> @JulienMalka in #283547 I proposed to use the `sort-key` to allow to reliably sort non-NixOS entries after NixOS ones, to avoid situations where a headless server boots into memtest86 for instance. It seems that the default entry is not always correctly selected otherwise and the first entry ends up being selected.
> 
> Do you see a way to support that in combination with this PR?

I'd advise supporting `sort key` for `memtest86` and `netbootxyz` maybe?

**Comment by r-vdp** [comment]
> > @JulienMalka in #283547 I proposed to use the `sort-key` to allow to reliably sort non-NixOS entries after NixOS ones, to avoid situations where a headless server boots into memtest86 for instance. It seems that the default entry is not always correctly selected otherwise and the first entry ends up being selected.
> > 
> > Do you see a way to support that in combination with this PR?
> 
> I'd advise supporting `sort key` for `memtest86` and `netbootxyz` maybe? 

Yeah that's what I did in that PR indeed. You mean that you wouldn't make the one for NixOS entries configurable?

I think that would be ok if the NixOS entries have a sort key somewhere in the middle of the alphabet so that people can easily sort other entries either before or after.

**Comment by JulienMalka** [comment]
> You mean that you wouldn't make the one for NixOS entries configurable?

Yes that is what I mean. I have to use the sort key to make the default generation be sorted first with the `boot-counting` feature enabled. I have documented why in https://github.com/NixOS/nixpkgs/pull/273062, but if you know of another solution I'm interested.

> I think that would be ok if the NixOS entries have a sort key somewhere in the middle of the alphabet so that people can easily sort other entries either before or after. 

After that PRs, nixos generations will have `sort-key` set to `default` or `non-default` but we can also prefix that with something like `nixos-default` and `nixos-non-default`, that way somebody can have other entries sorted before or after these ?

**Comment by r-vdp** [comment]
Yeah, we could even make the prefix configurable with `nixos` as the default value, to keep things flexible.

I didn't look in detail at this PR yet, but both options (configurable or not) sound good to me and will allow us to keep memtest at the bottom.

**Comment by JulienMalka** [comment]
@ofborg test systemd-boot

**Comment by JulienMalka** [comment]
@spiage if you have time would you please this PR?

**Comment by spiage** [comment]
@JulienMalka I have no issues with this commit. Are you asking about this?
```
$ sudo --preserve-env=NIXOS_INSTALL_BOOTLOADER -- nix-env -p /nix/var/nix/profiles/system --set /nix/store/2kpw7m28shk164p8k18lmyv58xwysc8m-nixos-system-nixos-24.05.20240127.9e9b136
$ sudo --preserve-env=NIXOS_INSTALL_BOOTLOADER -- systemd-run -E LOCALE_ARCHIVE -E NIXOS_INSTALL_BOOTLOADER --collect --no-ask-password --pty --quiet --same-dir --service-type=exec --unit=nixos-rebuild-switch-to-configuration --wait true
Using systemd-run to switch configuration.
$ sudo --preserve-env=NIXOS_INSTALL_BOOTLOADER -- systemd-run -E LOCALE_ARCHIVE -E NIXOS_INSTALL_BOOTLOADER --collect --no-ask-password --pty --quiet --same-dir --service-type=exec --unit=nixos-rebuild-switch-to-configuration --wait /nix/store/2kpw7m28shk164p8k18lmyv58xwysc8m-nixos-system-nixos-24.05.20240127.9e9b136/bin/switch-to-configuration switch
stopping the following units: accounts-daemon.service
activating the configuration...
setting up /etc...
reloading user units for spiage...
restarting sysinit-reactivation.target
reloading the following units: dbus.service
restarting the following units: polkit.service
starting the following units: accounts-daemon.service
the following new units were started: libvirtd.service
```
nixos-system-nixos: 24.05.20240127.c7a9d44 → 24.05.20240127.9e9b136

**Comment by spiage** [comment]
Yes, no issues, it was successful switch from master for my system
  #inputs.nixpkgs.url = "github:nixos/nixpkgs/master";
  inputs.nixpkgs.url = "github:nixos/nixpkgs/9e9b13658c4209f86f5ca5707babb17c7d1f8967";

**Comment by spiage** [comment]
I can reboot if you need )

**Review by RaitoBezarius — APPROVED** [review]
LGTM, we should test with people with old generations systemd boot, ideally without bootspec, etc, etc.

**Comment by JulienMalka** [comment]
Thanks, I did test with a machine with old generations, will also try to test with generations without bootspec and with strange boot entries. 
I'd encourage people with strange setups to test this as well.

**Comment by r-vdp** [comment]
@JulienMalka I reworked a bit the PR mentioned above (#283547) so that it can also deal properly with specialisations having different sort keys than their parent generations. This would also allow to control the fallback behaviour when using boot counting, since the specialisations will not necessarily be sorted right behind their parents anymore, but can be sorted separately (first all the top-level generations, and then all the specialisations).

When not setting a custom sort key for a specialisation, it inherits the sort key from its parent generation and we keep the current sorting behaviour.

Would it make sense to merge that PR first to introduce configurable sort keys, and then rebase this one on top? For this PR, I think we should then simply add a fixed suffix to determine what's the default entry (we add either `-default` or `-non-default` to every sort key). Like that, we can both have configurable sorting of entries and keep the functionality of this PR.

What do you think?

**Comment by ElvishJerricco** [comment]
I never got around to reviewing the original PR, so I guess here's my chance to review this one. I think the *code* looks good as is, but I question some of the design.

I rather don't like the way `sort-key` and `default` are handled here. IMO `sort-key` should be used to group profiles and specialisations, not default and non-default. I understand that the reason you've done it this way is because there's no way to specify the default in `loader.conf` such that a known-bad default won't be booted automatically. But I consider that a deficiency in systemd-boot (https://github.com/systemd/systemd/issues/31215) that should be fixed, rather than worked around.

On that note, I greatly dislike how specialisations of a known-bad generation are the first fallback tried. Using `sort-key` to group by specialisation would avoid this. i.e. The boot order would look like `Gen 2; Gen 1; Gen 2 Spec A; Gen 1 Spec A`.

So I sort of think we need:

1. @r-vdp's PR #283547 to group by specialisation,
2. A solution to https://github.com/systemd/systemd/issues/31215

Actually, I wonder if #283547 isn't even quite right. Rather than allowing the system derivation to choose its own `sort-key`, it seems to me like `sort-key` should just be equal to `nixos-${profile_name}-${specialisation_name}`

**Comment by JulienMalka** [comment]
> I understand that the reason you've done it this way is because there's no way to specify the default in `loader.conf` such that a known-bad default won't be booted automatically. But I consider that a deficiency in systemd-boot (https://github.com/systemd/systemd/issues/31215) that should be fixed, rather than worked around.

Thank you for opening the issue upstream, I think you are right calling this a deficiency on systemd-boot's side, will follow this carefully.  
> 
> On that note, I greatly dislike how specialisations of a known-bad generation are the first fallback tried. Using `sort-key` to group by specialisation would avoid this. i.e. The boot order would look like `Gen 2; Gen 1; Gen 2 Spec A; Gen 1 Spec A`.

I have considered this option, and have several issues with it:

- Firstly, you are making the assumption that specialisations and "mother" generation are close enough that if the mother generation is bad, the specialisations probably are too.  I don't think that it is true. I could do a NixOS configuration with specialisation just changing the kernel version for example, and in that case I'd happy that my different kernel versions are tried before going to another generation.

- Secondly, we could sort the specialisation after the normal generations, but we cannot have it both ways: the order of tries is the same as the boot menu. I think specialisations users want them sorted after the mother generations in the boot menue to be able to select them easily. 
These are the reasons that motivated my choice to leave the ordering of specialisations as is.

**Comment by ElvishJerricco** [comment]
Hm, I think those are moreso arguments that it should be user's choice. Because it depends on what the user has those specialisations for. For instance, I use specialisations so that every generation comes with a "rescue mode" with `rescue` in the kernel params (which *must* be done with specialisations when using lanzaboote). I would not want those specialisations to be tried until all the mother generations have been exhausted.

And if it's user's choice, we're back to https://github.com/NixOS/nixpkgs/pull/283547

**Comment by r-vdp** [comment]
I think sort-keys should be user-definable, the sorting of specialisations is one instance where people might have different preferences for how they are sorted, depending on what the specialisations are used for, but also because people might have other, non-nixos bootloader entries that they want to sort relatively to the nixos ones. So I think it makes sense to offer people this flexibility.

Ideally, boot counting would be completely orthogonal to this and would not need to rely on sort-keys.

**Comment by RaitoBezarius** [comment]
@ElvishJerricco I think fixing this in systemd will take time and a lot of users can still derive useful boot counting with the current state of things, do you see a reason to block this over the upstream fix?

**Comment by ElvishJerricco** [comment]
@RaitoBezarius Even without an upstream fix, I'd still like to see `sort-key` utilized better here. I suppose we can push that responsibility into https://github.com/NixOS/nixpkgs/pull/283547 though...

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius Even without an upstream fix, I'd still like to see `sort-key` utilized better here. I suppose we can push that responsibility into #283547 though...

That seems reasonable to me!

**Comment by JulienMalka** [comment]
I have interfaced this PR with https://github.com/NixOS/nixpkgs/pull/283547. Now that `sort-key` is user defined, we have to dynamically compute a suitable `sort-key` to display the default generation first.

**Comment by JulienMalka** [comment]
@ofborg test systemd-boot

File: nixos/modules/system/boot/loader/systemd-boot/boot-counting.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Automatic boot assessment with systemd-boot {#sec-automatic-boot-assessment}
+
+## Overview {#sec-automatic-boot-assessment-overview}
+
+Automatic boot assessment (or boot-counting) is a feature of `systemd-boot` that allows for automatically detecting invalid boot entries.
+When the feature is active, each boot entry has an associated counter with a user defined number of trials. Whenever `system-boot` boots an entry, its counter is decreased by one, ultimately being marked as *bad* if the counter ever reaches zero. However, if an entry is successfully booted, systemd will permanently mark it as *good* and remove the counter altogether. Whenever an entry is marked as *bad*, it is sorted last in the `systemd-boot` menu.
```
Comment (fd) [inline nixos/modules/system/boot/loader/systemd-boot/boot-counting.md:6]: There's a small typo here. `system-boot` 👉 `systemd-boot`

Comment (JulienMalka) [reply nixos/modules/system/boot/loader/systemd-boot/boot-counting.md]: Thanks

File: nixos/modules/system/boot/systemd.nix [hunk 2]
```diff
@@ -102,6 +102,10 @@ let
       "systemd-rfkill.service"
       "systemd-rfkill.socket"
 
+      # Boot counting
+      "boot-complete.target"
+    ] ++ lib.optional config.boot.loader.systemd-boot.bootCounting.enable "systemd-bless-boot.service" ++ [
```
Comment (arianvp) [inline nixos/modules/system/boot/systemd.nix:107]: Don't we also need `systemd-boot-check-no-failures.service` when `bootCounting.enable = true` ?  Or at least have an option to enable it?

Comment (JulienMalka) [reply nixos/modules/system/boot/systemd.nix:107]: > Note that due the simple nature of this check this service is probably not suitable for deployment in most scenarios. It is primarily useful only as example for developing more fine-grained checks to order before boot-complete.target.

This tells me we should maybe only add it as an option, yes.

**Comment by fd** [comment]
@JulienMalka is there anything blocking or can we get this PR merged?

**Comment by JulienMalka** [comment]
No really, especially now that branch-off has happened. I will rerun a test phase on real hardware and merge it next week if no further remarks.

File: nixos/tests/systemd-boot.nix [hunk 3]
```diff
@@ -93,7 +95,7 @@ in
       machine.wait_for_unit("multi-user.target")
 
       machine.succeed("test -e /boot/loader/entries/nixos-generation-1.conf")
-      machine.succeed("grep 'sort-key nixos' /boot/loader/entries/nixos-generation-1.conf")
+      machine.succeed("grep 'sort-key nixor-default' /boot/loader/entries/nixos-generation-1.conf")
```
Comment (glukki) [inline nixos/tests/systemd-boot.nix:98]: `nixoR-default` — typo?

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix [hunk 4]
```diff
@@ -313,6 +318,15 @@ in {
       '';
     };
 
+    bootCounting = {
+      enable = mkEnableOption (lib.mdDoc "automatic boot assessment");
+      trials = mkOption {
```
Comment (nikstur) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:323]: This should be called "tries" in line with the upstream documentation. E.g. kernel-install looks in `/etc/kernel/tries`.

```suggestion
      tries = mkOption {
```

File: nixos/modules/system/boot/systemd.nix (hunk 2 above)
Comment (nikstur) [reply nixos/modules/system/boot/systemd.nix:107]: `systemd-boot-check-no-failures.service` is probably way too strict and will not be useful in a generic system.

File: nixos/tests/systemd-boot.nix (hunk 3 above)
Comment (Mic92) [reply nixos/tests/systemd-boot.nix:98]: Looks like this is on purpose:

```
[root@bernie:~]# cat /boot/loader/entries/nixos-generation-208.conf
title NixOS
version Generation 208 NixOS Vicuna 24.11.20240725.3d0bda0 (Linux 6.9.10), built on 2024-07-25
linux /EFI/nixos/zdyjg9hmh9xrssl4bk55ws79vr9clh1y-linux-6.9.10-bzImage.efi
initrd     /EFI/nixos/z61f18hhpavf7xqxvgal3gh1qy40bbf7-initrd-linux-6.9.10-initrd.efi
options init=/nix/store/5i8knq7i2dp7q2kacjwr30f047kxvrnz-nixos-system-bernie-24.11.20240725.3d0bda0/init nohibernate root=fstab splash loglevel=4
machine-id 893c3816533c45ac939653b7816f8a18
sort-key nixos
[root@bernie:~]#

[root@bernie:~]# cat /boot/loader/entries/nixos-generation-209+3.conf
title NixOS
version Generation 209 NixOS Vicuna 24.11.20240725.3d0bda0 (Linux 6.9.10), built on 2024-07-25
linux /EFI/nixos/zdyjg9hmh9xrssl4bk55ws79vr9clh1y-linux-6.9.10-bzImage.efi
initrd   /EFI/nixos/z61f18hhpavf7xqxvgal3gh1qy40bbf7-initrd-linux-6.9.10-initrd.efi
options init=/nix/store/bnrbgjbbzb8a4kb27v2ycs9p033ar88i-nixos-system-bernie-24.11.20240725.3d0bda0/init nohibernate root=fstab splash loglevel=4
machine-id 893c3816533c45ac939653b7816f8a18
sort-key nixor-default
```

r < s

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix (hunk 4 above)
Comment (Mic92) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:323]: Done in https://github.com/NixOS/nixpkgs/pull/330017

**Comment by Mic92** [comment]
@JulienMalka I rebased and tested your code in: https://github.com/NixOS/nixpkgs/pull/330017

File: nixos/tests/systemd-boot.nix (hunk 3 above)
Comment (eclairevoyant) [reply nixos/tests/systemd-boot.nix:98]: should be commented then?

Comment (Mic92) [reply nixos/tests/systemd-boot.nix:98]: See https://github.com/NixOS/nixpkgs/commit/10f4b230b7a46f70444d5fe9036e9de490fb5e4d#diff-14341d580318ebe4f2ce22e4fc94c02f6a56229cdc7ae939728628a47b9e6b39R116
I also commented it in the test.

**Comment by JulienMalka** [comment]
Thank you @Mic92, closing

**Pull request closed without merging**